core/blockchain.go

```diff
@@ -1272,6 +1288,156 @@ func (bc *BlockChain) WriteBlockWithState(
 	return CanonStatTy, nil
 }
 
+func (bc *BlockChain) pruneBeaconChain(db ethdb.Database, maxBlockNum uint64, maxEpoch *big.Int) {
+	// check is start
+	if atomic.CompareAndSwapInt32(&bc.isPruneBeaconChainRunning, 0, 1) {
```

Comment:
Can you implement the same logic with just Mutex? It's simpler to understand and less error-prone.

Reply:
This equal `TryLock`, golang 'sync/mutex' package currently no this method, We can import other packages to solve this problem, but I don’t think it is necessary.

If we use `Lock` method, If the processing is slow, a waiting goroutine will create, this may cause memory leaks

Reply:
You may make this a function, e.g. `isPruning`. Also works for defer

Reply:
@JackyWYX Good idea, I optimized it